### PR TITLE
Avoid reshuffling when numOfShard is 0

### PR DESCRIPTION
**Pull request by mrkm4ntr**
This resolves part of #1938.

### CODE REVIEW

File: scio-elasticsearch/es6/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java [hunk 1]
```diff
@@ -367,30 +365,42 @@ public PDone expand(final PCollection<T> input) {
...
-                    new ElasticsearchWriter<>(
-                        clusterName,
-                        servers,
-                        maxBulkRequestSize,
-                        maxBulkRequestBytes,
-                        toDocWriteRequests,
-                        error,
-                        maxRetries,
-                        retryPause)));
+        if (numOfShard == 1) {
```
Comment (nevillelyh) [inline scio-elasticsearch/es6/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:368]: Should this be `0` instead? If `numOfShard == 1` we explicitly want to shuffle everything into a single shard? `0` is also consistent with file-based IOs, i.e. do not reshuffle, write each bundle as is.

Comment (mrkm4ntr) [reply scio-elasticsearch/es6/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java]: Addressed.

**Review by regadas — CHANGES_REQUESTED** [review]
@mrkm4ntr Thank you for addressing this! Overall looks good. Left some minor comments.

File: scio-elasticsearch/es6/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java [hunk 2]
```diff
@@ -498,56 +601,65 @@ private void flush(BulkRequest chunk) throws Exception {
...
+    private static final String RETRY_ATTEMPT_LOG =
+            "Error writing to Elasticsearch. Retry attempt[%d]";
+    private static final String RETRY_FAILED_LOG =
+            "Error writing to ES after %d attempt(s). No more attempts allowed";
 
-          if (exception != null) {
-            throw new Exception(String.format(RETRY_FAILED_LOG, attempt), exception);
+    private static ProcessFunction<BulkRequest, BulkResponse> retry(
+            final ProcessFunction<BulkRequest, BulkResponse> requestFn,
+            final FluentBackoff backoffConfig, final Logger logger) {
```
[the remark this answers is not in the document]
Comment (mrkm4ntr) [reply scio-elasticsearch/es6/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java]: This becomes the method of `Write` class. So I use Logger of `Write`.

File: scio-elasticsearch/es6/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java [hunk 3]
```diff
@@ -498,56 +601,65 @@ private void flush(BulkRequest chunk) throws Exception {
...
-
         try {
           requestFn.apply(chunk);
         } catch (Exception e) {
           retryFn.apply(chunk);
         }
       }
+    }
 
-      private static ProcessFunction<BulkRequest, BulkResponse> request(
```
[the remark this answers is not in the document]
Comment (mrkm4ntr) [reply scio-elasticsearch/es6/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:511]: Addressed.

**Comment by mrkm4ntr** [comment]
@regadas @nevillelyh PTAL.

**Review by regadas — APPROVED** [review]
Thank you @mrkm4ntr

**Pull request merged**